src/hotspot/share/gc/parallel/psParallelCompact.cpp

```diff
@@ -1256,16 +1256,28 @@ void PSParallelCompact::summarize_spaces_quick()
 }
 
 void PSParallelCompact::fill_dense_prefix_end(SpaceId id) {
-  // Since both markword and klass takes 1 heap word, the min-obj-size is 2
-  // heap words.
-  // If min-fill-size decreases to 1, this whole method becomes redundant.
-  assert(CollectedHeap::min_fill_size() == 2, "inv");
+  // Comparing two sizes to decide if filling is required:
+  //
+  // The size of the filler (min-obj-size) is 2 heap words with the default
+  // MinObjAlignment, since both markword and klass takes 1 heap word.
+  //
+  // The size of the gap (if any) right before dense-prefix-end is
+  // MinObjAlignment.
+  //
+  // Need to fill in the gap only if it's smaller than min-obj-size, and the
+  // filler obj will extend to next region.
+
+  // Note: If min-fill-size decreases to 1, this whole method becomes redundant.
+  assert(CollectedHeap::min_fill_size() >= 2, "inv");
 #ifndef _LP64
-  // In 32-bit system, min-obj-alignment is >= 8 bytes, so the gap (if any)
-  // right before denses-prefix must be greater than min-fill-size; nothing to
-  // do.
+  // In 32-bit system, each heap word is 4 bytes, so MinObjAlignment == 2.
+  // The gap is always equal to min-fill-size, so nothing to do.
   return;
 #endif
+  if (MinObjAlignment > 1) {
+    return;
+  }
+  assert(CollectedHeap::min_fill_size() == 2, "inv");
```

Comment:
When `MinObjAlignment` is 8 or 16 (`MinObjAlignment` is 1 or 2), the `min_fill_size()` returns **2**. When `ObjectAlignmentInBytes` is larger than 16 (`MinObjAlignment` is larger then 2), the `min_fill_size()` returns a value **larger than 2**.

According to the first `assert` statement and the `if` statement, we can know that:
After the `if` statement, the `MinObjAlignment` must be `1`, so the second `assert` statement must be `true`. Could these two `assert` statement be simplified to one?

Reply:
> Could these two assert statement be simplified to one?

Could you provide a code snippet of how you see this being structured? It's unclear to me what your suggestion here is.

Reply:
> When `MinObjAlignment` is 8 or 16 (`MinObjAlignment` is 1 or 2),

Sorry, one typo. The first `MinObjAlignment` should be `ObjectAlignmentInBytes`. I fixed the previous comment.

> Could these two assert statement be simplified to one?

Actually, the second `assert` can be removed because we can deduce it according to the first `assert` and the `if` statement.

I provided such unclear guide just because I think you or others may have a better way to revise the code (instead of simply removing the second one).

Reply:
Concretely, when the first `assert` statement is `true`, we can know `MinObjAlignment >= 1`. Then the `if` statement excludes the `MinObjAlignment > 1`. So, only when `MinObjAlignment == 1`, the code can run after the `if` statement, then the `min_fill_size()` must be **2**.

```
  assert(CollectedHeap::min_fill_size() >= 2, "inv");
// <-- here, `MinObjAlignment >= 1`
  if (MinObjAlignment > 1) {   // exclude the `MinObjAlignment > 1`
    return;
  }
// <- only one condition: `MinObjAlignment == 1`
  assert(CollectedHeap::min_fill_size() == 2, "inv"); // <-- This assert must be true, because `MinObjAlignment == 1`.
```

Reply:
I see what you mean. The second assert is mostly for documentation purpose. Readers might not connect `min_fill_size` with `MinObjAlignment` right away.

The second assert can probably be removed if the `if-check` is written as `if (CollectedHeap::min_fill_size() > 2)`. I didn't go for that version, as it might introduce more assembly, but maybe the impact/overhead is negligible, if any.

Reply:
> The second assert is mostly for documentation purpose.

OK.